Hw jv array list laptop

Iskamele opened the pull request.

Iskamele committed:
for laptop

Iskamele committed:
Added class-level variables:
  - DEFAULT_CAPACITY: int = 10 - to limit the capacity of the elementData array during standard initialization;
  - elementData: T[] - typed array;
  - size: int - counter of the number of elements inside the elementData array;

Added ArrayList() constructor for standard initialization of array of elementData objects, with DEFAULT_CAPACITY capacity.

Added constructor ArrayList(int capacity): void to initialize an array of elementData objects of any capacity.
The method expects a positive number greater than 0, otherwise it throws an IllegalArgumentException.

Implemented add(T value): void method from the List interface.
Inside the method, the providingListCapacity() method is immediately called to ensure the capacity of the elementData array, after which the value T value is added to elementData at index "size".
After that, increment size.

Implemented add(T value, int index): void method from List interface.
Inside the method, the rangeCheckForAdd(index) (to check the validity of the index) and providingListCapacity() methods are first called.
Then, using the arraycopy() method from System, we shift all the elements to the right after index.
Then we add a new value to elementData[index].
And increment size.

Implemented addAll(List<T> list): void method from the List interface.
We call providingListCapacity() to ensure the capacity of the elementData array.
After that, we declare a loop that goes through the entire incoming list:
We initialize the typed variable newElement, to which we assign the i-th value of list and add newElement to elementData.

Implemented get(int index): T void method from List interface.
First of all, the validity of the index is checked using the checkIndex(index) method, after which we return the value from elementData by the valid index.

Implemented set(T value, int index): void method from the List interface.
First of all, the method checks the validity of the index using checkIndex(index).
After that, according to the valine index, the value of elementData is replaced with a new value.

Implemented remove(int index): T void method from List interface.
First of all, the method checks the validity of the index using checkIndex(index).
The typed variable oldValue is declared and initialized into which the value is passed by index from the elementData array.
We call the fastRemove(index) method, into which we pass the index of the element to be removed.
After that, we return oldValue.

Implemented remove(T element): T void method from List interface.
Inside, the index variable is declared and initialized with a value of -1 in case we do not find the element we are looking for.
Added a loop that loops through all elements of elementData:
If element == null and i-th elementData == null, then we have found the required element and enter its index into the index variable;
If element != null and element equals i-th elementData, then repeat the same steps.
After exiting the loop, we check if index == -1, then we throw the NoSuchElementException error.
If everything is good, we initialize the new variable oldValue with the element at the index from elementData.
And we call fastRemove(index) to remove the element from elementData.
After that return oldValue.

Implemented size(): int, which returns the value of the size variable.

Implemented isEmpty(): boolean which returns true if size == 0.

Added a providingListCapacity(): void method that checks, if size >= elementData.length, then the grow() method is called which expands elementData's capacity.

Added rangeCheckForAdd(int index): void method that checks if index is valid. If index < 0 or index > size, ArrayListIndexOutOfBoundsException is thrown.

Added a grow(): void method that calls the grow(int minNewCapacity) method, passing it size + 1.

Added grow(int minNewCapacity): void method that copies the logic of the original grow.
(I didn't write it from scratch, but I'll try to explain it the way I understand it):
grow is passed size + 1 as minNewCapacity variable.
Inside we initialize currentCapacity = elementData.length.
And we announce:
- newCapacity, which will store the result of the calculation after the loop;
- And bufferCellsRemain, which will store the difference between newCapacity and minNewCapacity.
A do-while loop is declared, which, if necessary, can run until the capacity of elementData is expanded to the required value.
Inside the loop:
- newCapacity is calculated, which increases by x1.5 times each time;
- bufferCellsRemain is calculated;
- and currentCapacity is passed the value of newCapacity.
Thus, we will have newCapacity at the output of sufficient length to store all elements in it (and until INTEGER MAX_VALUE separates us).
After exiting the loop, we declare a new typed array newElementData, which will be == the value of newCapacity in length.
And we copy the contents of elementData into it.
Then we assign the newElementData array to the elementData array.

Added checkIndex(int index): void method that checks if index is valid. If index < 0 or index >= size, then an ArrayListIndexOutOfBoundsException is thrown.

Added a fastRemove(int index): void method that, using the arraycopy() method from System, shifts elements to the left after index, replacing the removed value with the next one after it.
Then increment size.

Iskamele commented:
-electricity, I'll fix it in the morning

Iskamele committed:
Removed ArrayList(int capacity) constructor for successful Java CI build.

Rommelua commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,125 @@
+        return size == 0;
+    }
+
+    private void providingListCapacity() {
```
```suggestion
    private void extendCapacityIfListFull() {
```

Rommelua commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,125 @@
+        grow(size + 1);
+    }
+
+    private void grow(int minNewCapacity) {
```
You do not use minNewCapacity. You do not need loop in grow method.

Iskamele commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,125 @@
+        }
+    }
+
+    private void grow() {
```
This logic was taken from the original class.

Iskamele commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,125 @@
-        return null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (element == null && elementData[i] == null) {
+                index = i;
+            }
+            if (element != null && element.equals(elementData[i])) {
+                index = i;
+            }
```
It's awesome! Thank you, Leonid

Iskamele committed:
Changed remove(T element): T method.
Removed index and oldValue variables due to uselessness.
The if block has been simplified to immediately return the i-th element to the remove() method.
If the condition of the if block does not work, we throw a NoSuchElementException error.

Removed method void grow() { grow(size + 1); }.

Changed method grow(int minNewCapacity) void.
Now the method accepts nothing.
Removed bufferCellsRemain and minNewCapacity variables as they are no longer needed.
Removed the do-while loop as it is not needed in the current implementation.
The calculation of the newCapacity variable from the loop has been saved.
